### PR TITLE
Change default firmware paths, allow for override with module parameter

### PR DESCRIPTION
**Pull request by plbossart**
As discussed with firmware team, we want more flexibility in the location of firmware and topology files.

1. there is no reason why firmware and topology should be in the same location
2. we could have symlinks pointing a directory to a different release made of multiple files
3. we could even deal with firmware signed with multiple keys.
4. we can have OEM releases with non-public firmware.

Change the default directory from /lib/firmware/intel to /lib/firmware/intel/sof, and provide the means to override it. For example with an existing setup, the following command works fine:

modprobe sof-pci-dev fw_path="intel" tplg_path="intel"

Tested on Up2 only. I am pretty sure I broke the SPI support, @lyakh your feedback would be welcome.
I also believe i broke HDAudio support but I don't have a platform to test and @RanderWang is away.

### CODE REVIEW

File: sound/soc/sof/sof-pci-dev.c [hunk 1]
```diff
@@ -227,6 +252,33 @@ static int sof_pci_probe(struct pci_dev *pci,
...
+	/* alternate fw and tplg filenames ? */
+	if (fw_path)
+		path_prefix = fw_path;
+	else
+		path_prefix = sof_pdata->desc->default_fw_path;
+
+	fw_filename = devm_kasprintf(dev, GFP_KERNEL,
+				     "%s/%s",
+				     path_prefix,
+				     sof_pdata->machine->sof_fw_filename);
```
Comment (ranj063) [inline sound/soc/sof/sof-pci-dev.c]: @plbossart this will definitely break HDA platoforms as sof_pdata->machine will be NULL.

File: sound/soc/sof/sof-acpi-dev.c [hunk 2]
```diff
@@ -24,6 +24,14 @@
 /* platform specific devices */
 #include "intel/shim.h"
 
+static char *fw_path;
+module_param(fw_path, charp, 0444);
+MODULE_PARM_DESC(fw_path, "alternate path for SOF firmware.");
+
+static char *tplg_path;
+module_param(tplg_path, charp, 0444);
+MODULE_PARM_DESC(tplg_path, "alternate path for SOF topology.");
```
Comment (lyakh) [inline sound/soc/sof/sof-acpi-dev.c:33]: Wouldn't it be better to have these parameters in a common sof module instead of repeating them in each interface module? We don't expect to load several of them simultaneously and even if we did, letting them have different firmware and topology paths seems like being too kind to the user to me ;-)

Comment (plbossart) [reply sound/soc/sof/sof-acpi-dev.c:33]: I am not happy with the copy/paste here, but I couldn't think of anything better. soc-acpi-dev and soc-pci-dev are the top-level modules that will setup everything, I don't know how we can introduce a lower-level common module.

File: sound/soc/sof/sof-pci-dev.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/sof/sof-pci-dev.c]: ah yes, this entire part should be conditional the machine being detected.
That's really why I wanted the Up with HDMI only working so that we can test this path. I don't have access to any other platform.

Comment (plbossart) [reply sound/soc/sof/sof-pci-dev.c]: fixed and tested on Up2.

File: sound/soc/sof/sof-acpi-dev.c (hunk 2 above)
Comment (plbossart) [reply sound/soc/sof/sof-acpi-dev.c:33]: I ended-up moving the filename handling to lower levels, and deal with the module parameters only in sof-pci/acpi/spi-dev. that factors in more code and gives us more opportunities for quirks, e.g. for Hdaudio which works fine now.

**Comment by plbossart** [comment]
@lyakh @ranj063 please review the updates. I changed the way filenames are handled, and pass all the info in the platform data (prefix+filename). This works fine on Up2 with the topology file added in PR https://github.com/thesofproject/sof/pull/974

the string handling is a bit hackish, if you have a better way of changing hda-generic.tplg into hda-generic-idisp.tplg I am all ears.

**Comment by ranj063** [comment]
@plbossart how about using strtok? As an example:

char filename[100] = "sof-hda-generic.tplg";
char outfilename[100];
char s[2] = ".";
char *file, *ext;

file = strtok(filename, s);
ext = strtok(NULL, s);

sprintf(outfilename, "%s-idisp.%s", file, ext);

File: sound/soc/intel/common/soc-acpi-intel-byt-match.c [hunk 3]
```diff
@@ -197,8 +197,8 @@ struct snd_soc_acpi_mach  snd_soc_acpi_intel_baytrail_machines[] = {
 	{
 		.id = "10EC5682",
 		.drv_name = "sof_rt5682",
-		.sof_fw_filename = "intel/sof-byt.ri",
-		.sof_tplg_filename = "intel/sof-byt-rt5682.tplg",
+		.sof_fw_filename = "sof-byt.ri",
+		.sof_tplg_filename = "sof-byt-rt5682.tplg",
```
Comment (ranj063) [inline sound/soc/intel/common/soc-acpi-intel-byt-match.c:201]: @plbossart any particular reason you do this in a separate patch and not with the rest of the byt fixes? and similarly for the cnl rt5682 as well?

Comment (plbossart) [reply sound/soc/intel/common/soc-acpi-intel-byt-match.c:201]: yes. the first batch corresponds to what is already upstream and the second is still local to sof-dev. Splitting the two makes my life easier when pushing things upstream and merging back.

**Comment by plbossart** [comment]
> @plbossart how about using strtok? As an example:
> 
> char filename[100] = "sof-hda-generic.tplg";
> char outfilename[100];
> char s[2] = ".";
> char *file, *ext;
> 
> file = strtok(filename, s);
> ext = strtok(NULL, s);
> 
> sprintf(outfilename, "%s-idisp.%s", file, ext);

I must admit I had never heard of strtok. I'll give it a try, thanks for the suggestion.

**Comment by plbossart** [comment]
v3 pushed, improved string replacement with strsep as suggested by @ranj063 (strtok is not re-entrant)

File: sound/soc/sof/sof-acpi-dev.c [hunk 4]
```diff
@@ -64,8 +76,10 @@ static struct sof_dev_desc sof_acpi_baytrailcr_desc = {
 	.resindex_imr_base = 2,
 	.irqindex_host_ipc = 0,
 	.chip_info = &byt_chip_info,
-	.nocodec_fw_filename = "intel/sof-byt.ri",
-	.nocodec_tplg_filename = "intel/sof-byt-nocodec.tplg",
+	.default_fw_path = "intel/sof",
```
Comment (cujomalainey) [inline sound/soc/sof/sof-acpi-dev.c:79]: so these will go into a folder inside the intel folder?

Comment (plbossart) [reply sound/soc/sof/sof-acpi-dev.c:79]: by default yes. But you *could* always symlink intel/sof to another directory if you want to change the location of your firmware with secret-sauce, or use the kernel module parameter to specify another location. It's a lot more flexible than when the existing files had to be in /lib/firmware/intel

**Comment by ranj063** [comment]
looks good @plbossart. I'll do a test run on my minnowboard tonight and approve this.

**Comment by plbossart** [comment]
last call for comments, I will merge this tomorrow unless there is an objection
